arrow/src/array/data.rs

```diff
@@ -878,6 +898,250 @@ impl ArrayData {
+                    );
+                }
+
+                // check range actually is low -> high
+                if start_offset > end_offset {
+                    return Err(ArrowError::InvalidArgumentError(format!(
+                        "Offset invariant failure: non-monotonic offset at slot {}: {} > {}",
+                        i, start_offset, end_offset))
+                    );
+                }
+
+                Ok((i, start_offset..end_offset))
+            })
+            .try_for_each(|res: Result<(usize, Range<usize>)>| {
+                let (item_index, range) = res?;
+                validate(item_index, range)
+            })
+    }
+
+    /// Ensures that all strings formed by the offsets in buffers[0]
+    /// into buffers[1] are valid utf8 sequences
+    fn validate_utf8<T>(&self) -> Result<()>
+    where
+        T: ArrowNativeType + std::convert::TryInto<usize> + num::Num + std::fmt::Display,
+    {
+        let offset_buffer = &self.buffers[0];
+        let values_buffer = &self.buffers[1].as_slice();
+
+        self.validate_each_offset::<T, _>(
+            offset_buffer,
+            values_buffer.len(),
+            |string_index, range| {
+                std::str::from_utf8(&values_buffer[range.clone()]).map_err(|e| {
+                    ArrowError::InvalidArgumentError(format!(
+                        "Invalid UTF8 sequence at string index {} ({:?}): {}",
+                        string_index, range, e
+                    ))
+                })?;
+                Ok(())
+            },
+        )
+    }
+
+    /// Ensures that all offsets in buffers[0] into buffers[1] are
+    /// between `0` and `offset_limit`
+    fn validate_offsets_full<T>(&self, offset_limit: usize) -> Result<()>
+    where
+        T: ArrowNativeType + std::convert::TryInto<usize> + num::Num + std::fmt::Display,
+    {
+        let offset_buffer = &self.buffers[0];
+
+        self.validate_each_offset::<T, _>(
+            offset_buffer,
+            offset_limit,
+            |_string_index, _range| {
+                // No validation applied to each value, but the iteration
+                // itself applies bounds checking to each range
+                Ok(())
+            },
+        )
+    }
+
+    /// Validates that each value in self.buffers (typed as T)
+    /// is within the range [0, max_value], inclusive
+    fn check_bounds<T>(&self, max_value: i64) -> Result<()>
+    where
+        T: ArrowNativeType + std::convert::TryInto<i64> + num::Num + std::fmt::Display,
+    {
+        let required_len = self.len + self.offset;
+        let buffer = &self.buffers[0];
+
+        // This should have been checked as part of `validate()` prior
+        // to calling `validate_full()` but double check to be sure
+        assert!(buffer.len() / std::mem::size_of::<T>() >= required_len);
+
+        // Justification: buffer size was validated above
+        let indexes: &[T] = unsafe { &(buffer.typed_data::<T>()[self.offset..]) };
```

Comment:
I haven't done a full review yet, but I tried running our test suite against this branch and see a panic relating to this line. We do some weird things with buffers and it could be a bug in our code, but I think this slice needs to be

```suggestion
        let indexes: &[T] = unsafe { &(buffer.typed_data::<T>()[self.offset..self.offset+self.len]) };
```

The code that triggered the panic is a `null_if` kernel that calculates a new null bitmap but reuses the data buffer from the original array. When applying this kernel to a slice of an arrow array, the resulting validity buffer will be shorter than the data buffer and the `is_null` call a few lines below tries to access the validity buffer out of bounds.

Unrelated note, I think `typed_data` does not actually need to be unsafe since it checks the alignment requirements. The very similarly implemented `MutableBuffer::typed_data_mut` is not marked as unsafe. The safety notes mention bool as a special case, but that is no longer an `ArrowNativeType` since a while.

Reply:
Thank you @jhorstmann  -- I agree the indexes should be bounded to the length+offset rather than whatever is in the buffer. 

I also agree that `buffer.typed_data` doesn't need to be marked as `unsafe` -- I'll file a separate ticket / PR for that change.

Reply:
Added a test for this case in 76baca1811. Excellent spot @jhorstmann

Reply:
Filed https://github.com/apache/arrow-rs/issues/996